composer/callbacks/checkpoint_saver.py

```diff
@@ -0,0 +1,439 @@
+        |                        | :meth:`~composer.core.time.Timer.sample_in_epoch`.    |
+        +------------------------+-------------------------------------------------------+
+        | ``{token}``            | The total token count, as returned by                 |
+        |                        | :meth:`~composer.core.time.Timer.token`.              |
+        +------------------------+-------------------------------------------------------+
+        | ``{token_in_epoch}``   | The token count in the current epoch, as returned by  |
+        |                        | :meth:`~composer.core.time.Timer.token_in_epoch`.     |
+        +------------------------+-------------------------------------------------------+
+
+        The resulting checkpoint file name will be approximately
+        ``name_format_string.format(rank=..., local_rank=..., ...)``.
+
+        If :attr:`name_format_string` does not specify a file extension, ``.pt`` (if not using DeepSpeed)
+        or ``.tar`` (if using DeepSpeed) will be appended.
+
+        Consider the following example:
+
+        *   The global rank is ``0``.
+        *   The current epoch count is ``1``.
+        *   The current batch_count is ``42``.
+        *   DeepSpeed is not being used.
+
+        .. testsetup:: composer.callbacks.checkpoint_saver.CheckpointSaver.format_checkpoint_name
+
+            state = State(
+                train_dataloader=train_dataloader,
+                max_duration='1ep',
+                model=model,
+                rank_zero_seed=0,
+            )
+            state.timer._batch.value = 42
+            state.timer._epoch.value = 1
+        
+        .. doctest:: composer.callbacks.checkpoint_saver.CheckpointSaver.format_checkpoint_name
+
+            >>> CheckpointSaver.format_checkpoint_name(state, "ep{epoch}-ba{batch}/rank_{rank}")
+            'ep1-ba42/rank_0.pt'
+        """
+        checkpoint_name = name_format_string.format(
+            rank=dist.get_global_rank(),
+            local_rank=dist.get_local_rank(),
+            world_size=dist.get_world_size(),
+            local_world_size=dist.get_local_world_size(),
+            node_rank=dist.get_node_rank(),
+            epoch=int(state.timer.epoch),
+            batch=int(state.timer.batch),
+            batch_in_epoch=int(state.timer.batch_in_epoch),
+            sample=int(state.timer.sample),
+            sample_in_epoch=int(state.timer.sample_in_epoch),
+            token=int(state.timer.token),
+            token_in_epoch=int(state.timer.token_in_epoch),
+        )
+        if (not _is_pt(checkpoint_name)) and (not _is_archive(checkpoint_name)):
+            # missing an extension
+            if state.is_model_deepspeed:
+                checkpoint_name += ".tar"
+            else:
+                checkpoint_name += ".pt"
+
+        return checkpoint_name
+
+    def save_checkpoint(self, state: State, checkpoint_filename: str) -> None:
```

Comment:
This should be a standalone (e.g. static) method, and easy for others to call independently whenever they want to save a state.

Reply:
See `composer.utils.checkpoint:save_checkpoint`